src/zimpatch.cpp

```diff
@@ -39,265 +41,158 @@ std::string NumberToString(int number)
-        {
-            mainUrl=zim::writer::Url(tempString);
-        }
-
-        tempString=diff_archive.getMetadata("layoutaurl");
-        if(diff_archive.hasEntryByPath(tempString)
-         ||start_archive.hasEntryByPath(tempString))
-        {
-            layoutUrl=zim::writer::Url(tempString);
-        }
-
-        std::string tmp="";
-        std::vector<std::string> rdlist;
-        tempString=diff_archive.getMetadata("redirectlist");
-        for(unsigned int i=0;i<tempString.size();i++)
-        {
-            if(tempString[i]!='\n')
-            {
-                tmp+=tempString[i];
-            }
-            else
-            {
-                rdlist.push_back(tmp);
-                tmp="";
-            }
-        }
-        for(unsigned i=0;i<rdlist.size();i=i+2)
-        {
-            redirectList[rdlist[i]] = rdlist[i+1];
-        }
+  //Now adding new articles in file_2
+  for (unsigned int index =0; index<+diff_archive.getEntryCount(); index++) {
+    auto entry=diff_archive.getEntryByPath(index);
 
-        std::string dllist = diff_archive.getMetadata("dlist");
-        tmp="";
-        //Computing list of deleted articles and storing them in a vector.
-        for(unsigned int i=0; i<dllist.size(); i++)
-        {
-            if(dllist[i]!='\n')
-                tmp+=dllist[i];
-            else
-            {
-                delete_list.push_back(tmp);
-                tmp="";
-            }
-        }
-
-        //Process dlist.
-        std::cout<<"\nProcessing Delete list..\n"<<std::flush;
-        dlist.resize(start_archive.getEntryCount()+diff_archive.getEntryCount());
-        for(unsigned int i=0;i<dlist.size();i++)
-        {
-            dlist[i]=0;
-        }
-        for(unsigned int i=0;i<delete_list.size();i++)
-        {
-            //Deleted articles will always be present in the start_file only.
-            dlist[start_archive.getEntryByPath(delete_list[i]).getIndex()]=1;
-        }
-        delete_list.clear();
-        return;
+    //If the article is already in file_1, it has been added.
+    if(start_archive.hasEntryByPath(entry.getPath())||isAdditionalMetadata(entry.getPath())) {
+      continue;
     }
 
-    zim::Uuid getUuid() const
-    {
-        return fileUid;
-    }
-    zim::writer::Url getMainPage() const
-    {
-        return mainUrl;
-    }
-    zim::writer::Url getLayoutPage() const
-    {
-        return layoutUrl;
+    try {
+      // entry has been replace by a redirect in new zim file.
+      auto redirectPath = redirectList.at(entry.getPath());
+      zimCreator.addRedirection(entry.getPath(), entry.getTitle(), redirectPath);
+      continue;
+    } catch(...) {
+      if (entry.isRedirect()) {
+        zimCreator.addRedirection(entry.getPath(), entry.getTitle(), entry.getRedirectEntry().getPath());
+      } else {
+        auto tmpItem = std::shared_ptr<zim::writer::Item>(new CopyItem(entry.getItem()));
+        zimCreator.addItem(tmpItem);
+      }
     }
```

Comment:
Code duplication with a corresponding block in the previous loop. Looks like it can be converted into a free function taking only two parameters: `addEntry(zimCreator, entry)`

Reply:
Yes. But I really don't want to spend to much time on zimpatch/zimdiff.
They are tools created on long time ago and I disagree with the code itself.

I agree with the idea of having this kind of tools but I think this should be tackled at an another level. Something lower level working directly on cluster instead of article. And using another format that zim to store the patch file.

The code is here and we have to port it to new libzim api but probably nothing more (at least for now).